remme/tp/pub_key.py

```diff
@@ -61,95 +70,167 @@ def get_state_processor(self):
-                                                     backend=default_backend())
-        except ValueError:
-            raise InvalidTransaction('Cannot deserialize the provided public key: check if it is in PEM format')
-
-        try:
-            ehs_bytes = binascii.unhexlify(transaction_payload.entity_hash_signature)
-            eh_bytes = binascii.unhexlify(transaction_payload.entity_hash)
-        except binascii.Error:
-            LOGGER.debug(f'entity_hash_signature {transaction_payload.entity_hash_signature}')
-            LOGGER.debug(f'entity_hash {transaction_payload.entity_hash}')
-            raise InvalidTransaction('Entity hash or signature not a hex format')
-
+    @staticmethod
+    def _is_signature_valid(certificate_public_key, ehs_bytes, eh_bytes):
         # FIXME: For support PKCS1v15 and PSS
-        LOGGER.warn('HAZARD: Detecting padding for verification')
+        LOGGER.warning('HAZARD: Detecting padding for verification')
         sigerr = 0
         pkcs = padding.PKCS1v15()
         pss = padding.PSS(mgf=padding.MGF1(hashes.SHA512()), salt_length=padding.PSS.MAX_LENGTH)
         for _padding in (pkcs, pss):
             try:
-                cert_signer_pubkey.verify(ehs_bytes, eh_bytes, _padding, hashes.SHA512())
-                LOGGER.warn('HAZARD: Padding found: %s', _padding.name)
+                certificate_public_key.verify(ehs_bytes, eh_bytes, _padding, hashes.SHA512())
+                LOGGER.warning('HAZARD: Padding found: %s', _padding.name)
             except InvalidSignature:
                 sigerr += 1
 
         if sigerr == 2:
-            raise InvalidTransaction('Invalid signature')
+            return False
+
+        return True
+
+    @staticmethod
+    def _is_certificate_validity_exceeded(valid_from, valid_to):
+        """
+        Check if certificate validity exceeds the maximum value.
+
+        Certificate public key validity maximum value in one year (365 days).
+        """
+        valid_from = datetime.fromtimestamp(valid_from)
+        valid_to = datetime.fromtimestamp(valid_to)
+
+        if valid_to - valid_from > CERTIFICATE_PUBLIC_KEY_MAXIMUM_VALIDITY:
+            return False
+
+        return True
+
+    @staticmethod
+    def _charge_tokens_for_storing(context, address_from, address_to):
+        """
+        Send fixed tokens value from address that want to store certificate public key to node's storage address.
+        """
+        transfer_payload = TransferPayload()
+        transfer_payload.address_to = address_to
+        transfer_payload.value = PUB_KEY_STORE_PRICE
+
+        transfer_state = AccountHandler()._transfer_from_address(
+            context=context, address_from=address_from, transfer_payload=transfer_payload,
+        )
+
+        return transfer_state
+
+    def _store_pub_key(self, context, signer_public_key, new_public_key_payload):
+        """
+        Store certificate public key to the blockchain.
```

Comment:
And again :)
This can be not only a certificate, but also any other container.

Reply:
@eugene-babichenko, got you.